openedx/core/djangoapps/programs/utils.py

```diff
@@ -1,49 +1,105 @@
-    )
+def get_programs(user):
+    """Given a user, get programs from the Programs service.
 
+    Returned value is cached depending on user permissions. Staff users making requests
+    against Programs will receive unpublished programs, while regular users will only receive
+    published programs.
 
-def is_cache_enabled_for_programs():    # pylint: disable=invalid-name
-    """Returns a Boolean indicating whether responses from the Programs API
-    will be cached.
+    Arguments:
+        user (User): The user to authenticate as when requesting programs.
+
+    Returns:
+        list of dict, representing programs returned by the Programs service.
     """
-    return ProgramsApiConfig.current().is_cache_enabled
+    programs_config = ProgramsApiConfig.current()
+    cache_key = programs_config.get_cache_key(user)
+    no_programs = []
+
+    if not programs_config.enabled:
+        log.warning('Programs configuration is disabled.')
+        return no_programs
+
+    if programs_config.is_cache_enabled:
+        cached = cache.get(cache_key)
+        if cached:
+            return cached
+
+    try:
+        jwt = get_id_token(user, programs_config.OAUTH2_CLIENT_NAME)
+        api = EdxRestApiClient(programs_config.internal_api_url, jwt=jwt)
+    except Exception:  # pylint: disable=broad-except
+        log.exception('Failed to initialize the Programs API client.')
+        return no_programs
 
+    try:
+        response = api.programs.get()
+    except Exception:  # pylint: disable=broad-except
+        log.exception('Failed to retrieve programs from the Programs API.')
+        return no_programs
 
-def set_cached_programs_response(programs_data):
-    """ Set cache value for the programs data with specific ttl.
+    results = response.get('results', no_programs)
+
+    if programs_config.is_cache_enabled:
+        cache.set(cache_key, results, programs_config.cache_ttl)
+
+    return results
+
+
+def get_programs_for_dashboard(user, course_keys):
+    """Build a dictionary of programs, keyed by course.
+
+    Given a user and an iterable of course keys, find all the programs relevant
+    to the user's dashboard and return them in a dictionary keyed by course key.
 
     Arguments:
-        programs_data (dict): Programs data in dictionary format
+        user (User): The user to authenticate as when requesting programs.
+        course_keys (list): List of course keys representing the courses in which
+            the given user has active enrollments.
+
+    Returns:
+        dict, containing programs keyed by course. Empty if programs cannot be retrieved.
     """
-    cache.set(
-        ProgramsApiConfig.PROGRAMS_API_CACHE_KEY,
-        programs_data,
-        ProgramsApiConfig.current().cache_ttl
-    )
+    programs_config = ProgramsApiConfig.current()
+    course_programs = {}
+
+    if not programs_config.is_student_dashboard_enabled:
+        log.warning('Display of programs on the student dashboard is disabled.')
```

Comment:
now that we aren't checking the feature flag on the dashboard before calling this API, this warning will generate a lot of noise.

I'm actually thinking it would be better to restore the flag on the dashboard and leave the warning.